pulseapi/users/utils.py

```diff
@@ -34,6 +36,43 @@ def verify_recaptcha(response_token):
     return True
 
 
+__old_get_context_data = LoginView.get_context_data
+
+
+def augmented_get_context_data(self, **kwargs):
+    """
+    Patch allauth's LoginView.get_context_data class function
+    so that we can check for a recaptcha-related session value
+    if we're using recaptcha. Allauth only has post-processing
+    signals, so we're kind of left with monkey patching as the
+    only way to pre-process the login route.
+    """
+
+    if settings.USE_RECAPTCHA:
+        request = self.request
+        session = request.session
+
+        if 'recaptcha_token' not in session:
+            raise PermissionDenied()
+
+        session_token = session['recaptcha_token']
+        session['recaptcha_token'] = None
```

Comment:
clear the token after reading it so someone can't just set a valid session and then bulk-load the allauth route in a hundred tabs

Reply:
**Suggestion** This would probably make a good code comment  🙂